### PR TITLE
Refactor metrics for better cross algorithm reuse

### LINKED ISSUE #492: Refactor loss functions into a separate utils file
With the different possibilities like LK_Loss, KLDIV, CE loss, etc., pull these out of the specific training method core.py and make a utils file with loss functions

This would improve code organization and reusability by:
- Consolidating all loss function implementations in one place
- Making loss functions easier to discover and maintain
- Reducing clutter in the training method core.py file
- Enabling easier reuse across different training methods

### PR DESCRIPTION
**Pull request by fynnsu**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

Currently we aren't reusing a lot of code between different speculative algorithms for metric / loss calculation. This pr refactors the implementation so that parts can be shared. This also enables CE and KL_DIV loss for both eagle3 and dflash (whereas previously only one was implemented for each). 

Behavior should mostly remain the same for the two existing spec decoding algorithms. 

## Related Issue

Closes #492 

## Tests

Added some metrics tests.
E2E test: https://github.com/neuralmagic/llm-compressor-testing/actions/runs/25190304741

I have filled in:

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan/results, such as providing test command and pasting the results.
- [ ] (Optional) The necessary documentation update.
- [x] I (a human) have written or reviewed the code in this pr to the best of my ability.

### CODE REVIEW

File: src/speculators/models/metrics.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        cond_denom = prev_correct.sum()
+        # Update prev_correct in place
+        correct = torch.logical_and(prev_correct, correct, out=prev_correct)
+    if loss_mask is not None:
+        correct = torch.masked_select(correct, loss_mask.to(torch.bool))
+
+    correct_sum = correct.float().sum()
+    full_denom = correct.numel()
+
+    return correct_sum / (full_denom + 1e-5), correct_sum / (cond_denom + 1e-5)
```
Comment (shanjiaz) [inline src/speculators/models/metrics.py]: Can you save this magic number in a variable.

**Review by shanjiaz — APPROVED** [review]
e2e seems to be passing. Diff looks good to me! Maybe add a docstring in metrics.py or somewhere indicating we're passing logits instead of targets now?
